Assets/__Scripts/Map/BeatSaberMapV3.cs

```diff
@@ -0,0 +1,329 @@
+    {
+        DirectoryAndFile = other.DirectoryAndFile;
+        Time = other.Time;
+        Events = other.Events;
+        Notes = other.Notes;
+        Obstacles = other.Obstacles;
+
+        base.Waypoints = other.Waypoints;
+        BpmChanges = other.BpmChanges;
+        Bookmarks = other.Bookmarks;
+        CustomEvents = other.CustomEvents;
+        EnvEnhancements = other.EnvEnhancements;
+        CustomData = other.CustomData;
+        // do not ref mainnode, or it will exist duplicate data.
+    }
+
+    public override bool Save()
+    {
+        if (!Settings.Instance.Load_MapV3)
+        {
+            return base.Save();
+        }
+        try
+        {
+            /*
+             * LISTS
+             */
+
+            //Just in case, I'm moving this up here
+            Thread.CurrentThread.CurrentCulture = CultureInfo.InvariantCulture;
+            Thread.CurrentThread.CurrentUICulture = CultureInfo.InvariantCulture;
+
+            if (MainNode is null) MainNode = new JSONObject();
+
+            MainNode["version"] = Version;
+            ParseBaseNoteToV3();
+            /// official nodes
+            var bpmEvents = new JSONArray();
+            foreach (var b in BpmEvents) bpmEvents.Add(b.ConvertToJson());
+
+            var rotationEvents = new JSONArray();
+            foreach (var b in RotationEvents) RotationEvents.Add(b);
+
+            var colorNotes = new JSONArray();
+            foreach (var n in ColorNotes) colorNotes.Add(n.ConvertToJson());
+
+            var bombNotes = new JSONArray();
+            foreach (var b in BombNotes) bombNotes.Add(b.ConvertToJson());
+
+            var obstacles = new JSONArray();
+            foreach (var o in ObstaclesV3) obstacles.Add(o.ConvertToJson());
+
+            var sliders = new JSONArray();
+            foreach (var s in Sliders) sliders.Add(s.ConvertToJson());
+
+            var chains = new JSONArray();
+            foreach (var c in Chains) chains.Add(c.ConvertToJson());
+
+            var waypoints = new JSONArray(); // TODO: Add formal support
+            foreach (var w in Waypoints) waypoints.Add(w);
+
+            var basicBeatmapEvents = new JSONArray();
+            foreach (var b in BasicBeatmapEvents) basicBeatmapEvents.Add(b.ConvertToJson());
+
+            var colorBoostBeatmapEvents = new JSONArray();
+            foreach (var c in ColorBoostBeatmapEvents) colorBoostBeatmapEvents.Add(c.ConvertToJson());
+
+            var lightColorEventBoxGroups = new JSONArray();
+            foreach (var l in LightColorEventBoxGroups) lightColorEventBoxGroups.Add(l);
+
+            var lightRotationEventBoxGroups = new JSONArray();
+            foreach (var l in lightRotationEventBoxGroups) lightRotationEventBoxGroups.Add(l);
+
+            var basicEventTypesWithKeywords = new JSONArray();
+            foreach (var b in BasicEventTypesWithKeywords) basicEventTypesWithKeywords.Add(b);
+
+            MainNode["bpmEvents"] = CleanupArray(bpmEvents, "b");
```

Comment:
Let's not use Official BPM Changes. They have been infamously buggy, and from @JD-The-65th's comment, likely still are to this day. For the time being, let's continue to put them in Custom Data.

I asked Aeroluna a few days ago about how Custom Data will be stored in Beatmap v3, and he told me this:
- Custom data in `Info.dat` will still be stored in `_customData`
- Custom data in the root of the Difficulty (like our BPM Changes) will be stored in `customData`
- Custom data per object (Chroma event data, NE data, etc.) will be stored in `cd`

Reply:
Also, if you want to experience the effects of official BPM changes yourself, look no further than lawless Africa https://beatsaver.com/maps/1d233

Reply:
Thanks for the suggestion. Then I think I could leave bpm change still for customdata to parse. I don't have to take care of it(However I'm not sure what effect it will be with slider and chain).
Besides, for customdata, I think I need to wait for a chromma supprot and test its functionality,